client/session.go

```diff
@@ -1548,29 +1590,164 @@ func (s *session) streamBlocksMetadataFromPeer(
 	return nil
 }
 
+// streamBlocksMetadataFromPeerV2 has several heap allocated anonymous
+// function, however, they're only allocated once per peer/shard combination
+// for the entire peer bootstrapping process so performance is acceptable
+func (s *session) streamBlocksMetadataFromPeerV2(
+	namespace ts.ID,
+	shard uint32,
+	peer peer,
+	start, end time.Time,
+	metadataCh chan<- blocksMetadata,
+	progress *streamFromPeersMetrics,
+) error {
+	var (
+		pageToken []byte
+		retrier   = xretry.NewRetrier(xretry.NewOptions().
```

Comment:
Ok I took a quick glance through the retrier code and it looks thread safe so this should be fine

Reply:
actually there is already a fetch retrier in the struct, I'll just reuse that

Reply:
Ok so that retrier had some defaults that were different than what we hard-coded before, so to be safe I created a streamBlocksRetrier in the options which has the same defaults as what used to be hard-coded